### PR TITLE
Get prime factors from a number

### PR DESCRIPTION
**Pull request by FGS-TBittner**
Functionality completed to get a list of all prime factors of a number

### CODE REVIEW

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        [Test]
+        public void GetPrimeFactors_Given1_ReturnsEmptyList()
+        {
+            CollectionAssert.IsEmpty(_factorizer.GetPrimeFactors(1));
+        }
+
+        [TestCase(6, new int[] { 2, 3 })]
+        [TestCase(100, new int[] { 2, 2, 5, 5 })]
+        [TestCase(2310, new int[] { 2, 3, 5, 7, 11 })]
+        [TestCase(398, new int[] { 2, 199 })]
```
Comment (fgs-dbudwin) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: There is a preferred way to write this.  In NUnit you can use the `TestCaseSource` attribute.  Then you can create a method with all of the current test case scenarios being `yield return` statements.  This will be true for all your tests with multiple test cases.

File: PrimeFactors/PrimeNumberFactorizer.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            var factor = LOWEST_PRIME_NUMBER;
+
+            while (factor < num && num % factor != 0)
+                factor++;
+
+            return factor;
+        }
+
+        private int RemoveLowestFactor(int num)
```
Comment (fgs-dbudwin) [inline PrimeFactors/PrimeNumberFactorizer.cs]: The name of this function is misleading, it doesn't actually remove anything.  It appears to calculate the lower prime factor to remove if I understand correctly.

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace PrimeFactors.Tests
+{
+    [TestFixture]
+    public class PrimeNumberFactorizerTests
+    {
+        private PrimeNumberFactorizer _factorizer;
+
+        [SetUp]
+        public void Init()
```
Comment (fgs-dbudwin) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: Please rename to `Setup()`

File: PrimeFactors/PrimeNumberFactorizer.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace PrimeFactors
+{
+    public class PrimeNumberFactorizer
+    {
+        private const int LOWEST_PRIME_NUMBER = 2;
+
+        public List<int> GetPrimeFactors(int num)
```
Comment (fgs-dbudwin) [inline PrimeFactors/PrimeNumberFactorizer.cs]: It would be best practice to return an `IEnumerable<int>` from here.

Also since negative numbers cannot be prime, we should be careful about our types.  Let's actually make this `IEnumerable<uint>`.

File: PrimeFactors/PrimeNumberFactorizer.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+            while (num >= LOWEST_PRIME_NUMBER)
+            {
+                primeFactors.Add(GetLowestFactor(num));
+                num = RemoveLowestFactor(num);
+            }
+            return primeFactors;
+        }
+
+        private int GetLowestFactor(int num)
```
Comment (fgs-dbudwin) [inline PrimeFactors/PrimeNumberFactorizer.cs]: Ditto, `uint`

File: PrimeFactors/PrimeNumberFactorizer.cs (hunk 2 above)
Comment (fgs-dbudwin) [inline PrimeFactors/PrimeNumberFactorizer.cs]: Ditto, `uint`

File: PrimeFactors/PrimeNumberFactorizer.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace PrimeFactors
+{
+    public class PrimeNumberFactorizer
+    {
+        private const int LOWEST_PRIME_NUMBER = 2;
```
Comment (fgs-dbudwin) [inline PrimeFactors/PrimeNumberFactorizer.cs]: Ditto, `uint`

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs (hunk 1 above)
Comment (FGS-TBittner) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: When declaring the methods for `TestCaseSource` where is the preferred place of declaration: top of `class`, bottom of `class`, after the `Test` function that uses it, or in a completely different `class` of it's own?

Comment (fgs-dbudwin) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: At the bottom of the class is fine.  We like to keep the test methods at the top for visibility.

File: PrimeFactors/PrimeNumberFactorizer.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            var factor = LOWEST_PRIME_NUMBER;
+
+            while (factor < num && num % factor != 0)
+                factor++;
+
+            return factor;
+        }
+
+        private uint GetNumberWithLowestPrimeFactorRemoved(uint num)
```
Comment (fgs-dbudwin) [inline PrimeFactors/PrimeNumberFactorizer.cs]: I still have an issue with this name...if I see `Remove[d]` in the function name, I expect to see something removed from a collection or a database or a delete.

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            CollectionAssert.IsEmpty(_factorizer.GetPrimeFactors(0));
+        }
+
+        [Test]
+        public void GetPrimeFactors_Given1_ReturnsEmptyList()
+        {
+            CollectionAssert.IsEmpty(_factorizer.GetPrimeFactors(1));
+        }
+
+        [TestCaseSource("NonPrimeNumbersAndExpectedFactors")]
```
Comment (fgs-dbudwin) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: Doing it this way, while working, isn't very maintainable since it can't be refactored easily.  I would suggest you change this to `[TestCaseSource(nameof(NonPrimeNumbersAndExpectedFactors))]` that way if you were to rename `NonPrimeNumbersAndExpectedFactors`, it will be updated here too.

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        public void GetPrimeFactors_GivenPrimeNumberSquared_ReturnsSquareRootOfNumberDuplicated(uint num, uint sqrt)
+        {
+            var factors = new List<uint>();
+            factors.Add(sqrt);
+            factors.Add(sqrt);
+
+            CollectionAssert.AreEqual(factors, _factorizer.GetPrimeFactors(num));
+        }
+
+        public static IEnumerable<object> NonPrimeNumbersAndExpectedFactors()
```
Comment (fgs-dbudwin) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: Eek!  Probably not the best use of `object`.  You may need to make a `private` class that represents `object` instead.

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs (hunk 8 above)
Comment (FGS-TBittner) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: Converting the `TestCaseSource` functions to private classes would result in my `Test` file having a few private classes contained within it. Is that normal, or should those be in separate `class` files in the `Test` project?

Comment (FGS-TBittner) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: I miss understood `nameof` vs `typeof` please disregard

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs (hunk 9 above)
Comment (fgs-pwright) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: Instead of declaring this method an `IEnumerable<object>`, you could use NUnit's `TestCaseData` class, and `yield return` a `TestCaseData` instance for each of these tests you want to run.  There's some documentation for it [here](https://github.com/nunit/docs/wiki/TestCaseData).

File: PrimeFactors/PrimeNumberFactorizer.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        private const uint LOWEST_PRIME_NUMBER = 2;
+
+        public IEnumerable<uint> GetPrimeFactors(uint num)
+        {
+            var primeFactors = new List<uint>();
+
+            while (num >= LOWEST_PRIME_NUMBER)
+            {
+                primeFactors.Add(GetLowestPrimeFactor(num));
+                num = GetNumberWithLowestPrimeFactorRemoved(num);
```
Comment (fgs-pwright) [inline PrimeFactors/PrimeNumberFactorizer.cs]: This does what you want it to, but you end up running `GetLowestPrimeFactor()` twice in each pass through the loop, with the same parameter passed to it.  Is there any way you can optimize that?

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs (hunk 9 above)
Comment (fgs-dbudwin) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: Also, these should be `private`

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        {
+            yield return new object[] { 6U, new uint[] { 2U, 3U } };
```
Comment (fgs-dbudwin) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: Why did you add `U` to all of these?  Shouldn't it know that it's unsigned because of the type of the array?

Comment (FGS-TBittner) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: I got some conversion error when I initially changed everything to `uint`, but the code is different now it may be able to handle it

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs (hunk 9 above)
Comment (fgs-pwright) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: n.b., unlike in the docs, we will typically declare the `IEnumerable<TestCaseData>` methods within the class containing our tests.  You shouldn't need to change your `[TestCaseSource(...)]` attribute, other than Drew's note about it above.

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs [hunk 12]
```diff
@@ -31,13 +31,13 @@ public void GetPrimeFactors_Given1_ReturnsEmptyList()
             CollectionAssert.IsEmpty(_factorizer.GetPrimeFactors(1));
         }
 
-        [TestCaseSource("NonPrimeNumbersAndExpectedFactors")]
-        public void GetPrimeFactors_GivenNonPrimeNumber_ReturnsExpectedFactors(uint num, IEnumerable<uint> factors)
+        [TestCaseSource(nameof(NonPrimeNumbersAndExpectedFactors))]
+        public void GetPrimeFactors_GivenNonPrimeNumber_ReturnsExpectedFactors(uint num, IEnumerable factors)
```
Comment (fgs-dbudwin) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: Why did you remove the `<uint>`?

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs [hunk 13]
```diff
@@ -56,29 +56,29 @@ public void GetPrimeFactors_GivenPrimeNumberSquared_ReturnsSquareRootOfNumberDup
...
-            yield return 199U;
+            yield return 2;
+            yield return 3;
+            yield return 5;
+            yield return 151;
+            yield return 199;
         }
 
-        public static IEnumerable<object> PrimeNumberSquaredAndSquareRoot()
+        private static IEnumerable<uint[]> PrimeNumberSquaredAndSquareRoot()
```
Comment (fgs-dbudwin) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: I would advise against returning an `IEnumerable` with an array.  I think you should follow @fgs-pwright advice and use `TestCaseData` here as well.

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs (hunk 12 above)
Comment (FGS-TBittner) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: I had attempted to replace the parameters with a private class, and when I switched to using `TestCaseData` I had forgotten to add the `<uint>` to `IEnumerable` when I replaced the parameters

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs [hunk 14]
```diff
@@ -73,12 +73,12 @@ private static IEnumerable<uint> PrimeNumbers()
...
         }
 
-        private static IEnumerable<uint[]> PrimeNumberSquaredAndSquareRoot()
+        private static IEnumerable<TestCaseData> PrimeNumberSquaredAndSquareRoot()
         {
-            yield return new uint[] { 4, 2 };
-            yield return new uint[] { 49, 7 };
-            yield return new uint[] { 121, 11 };
-            yield return new uint[] { 38809, 197 };
+            yield return new TestCaseData(4U, 2U);
```
Comment (fgs-dbudwin) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: Why did you put the `U`'s back?

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+            var factors = new List<uint>();
+            factors.Add(sqrt);
+            factors.Add(sqrt);
+
+            CollectionAssert.AreEqual(factors, _factorizer.GetPrimeFactors(num));
+        }
+
+        private static IEnumerable<TestCaseData> NonPrimeNumbersAndExpectedFactors()
+        {
+            yield return new TestCaseData(6U, new uint[] { 2, 3 });
```
Comment (fgs-dbudwin) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: Same here, why did you put the `U` back.  These should pass just fine to a parameter that's a `uint`.

**Comment by fgs-dbudwin** [comment]
http://onepeggenius.com/wp-content/uploads/2009/09/the-u-is-back.gif

File: PrimeFactors.Tests/PrimeNumberFactorizerTests.cs (hunk 15 above)
Comment (FGS-TBittner) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: The `U` had to be specified in the constructor for `TestCaseData` otherwise I get the error `Object of type 'System.Int32' cannot be converted to type 'System.UInt32'` when NUnit goes to run tests

Comment (fgs-dbudwin) [inline PrimeFactors.Tests/PrimeNumberFactorizerTests.cs]: This works on my machine:

```
private static IEnumerable<TestCaseData> PrimeNumberSquaredAndSquareRoot()
        {
            yield return new TestCaseData(4, 2);
            yield return new TestCaseData(49, 7);
            yield return new TestCaseData(121, 11);
            yield return new TestCaseData(38809, 197);
        }
```

**Comment by fgs-pwright** [comment]
Nope.  Nice work, Taylor!

**Comment by fgs-dbudwin** [comment]
:gopher: :squash:

**Comment by fgs-dbudwin** [comment]
https://cdn.drawception.com/images/panels/2015/7-19/ZY2Gd956M4-12.png
